### ISSUE TITLE
google.api_core.exceptions.ServiceUnavailable: 503 Deadline Exceeded

### ISSUE THREAD

**Issue by praharshbhatt** [googleapis/python-firestore#4]
#### Environment details
OS: Win 10
Python Version: 2.7
google-cloud-firestore

I get the deadline exceeded error when I try to do any operation on the database.

#### Code example

```
from google.cloud import firestore

def setCredentials():
    from google.oauth2 import service_account
    credentials = service_account.Credentials.from_service_account_file("fall detection-c5f059887a9d.json")
    return credentials


def main():
    db = firestore.Client(credentials=setCredentials())

    # Create a new collection and a document
    doc_ref = db.collection(u'users').document(u'1')
    doc_ref.set({
        u'first': u'Ada',
        u'last': u'Lovelace',
        u'born': 1815
    })
    print("ok")


if __name__ == "__main__":
    main()

```

#### Stack trace
```
"D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\Scripts\python.exe" "D:\Apps\JetBrains\PyCharm Community Edition 2019.1.3\helpers\pydev\pydevd.py" --multiproc --qt-support=auto --client 127.0.0.1 --port 55450 --file "D:/Multiverse_Projects/Final year project- Fall Detection/video-fall-detection/cloud-firestore-test.py"
pydev debugger: process 15592 is connecting

Connected to pydev debugger (build 191.7479.30)
D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\urllib3\util\ssl_.py:380: SNIMissingWarning: An HTTPS request has been made, but the SNI (Server Name Indication) extension to TLS is not available on this platform. This may cause the server to present an incorrect TLS certificate, which can cause validation failures. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  SNIMissingWarning,
D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\urllib3\util\ssl_.py:139: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecurePlatformWarning,
No handlers could be found for logger "concurrent.futures"
Traceback (most recent call last):
  File "D:\Apps\JetBrains\PyCharm Community Edition 2019.1.3\helpers\pydev\pydevd.py", line 1758, in <module>
    main()
  File "D:\Apps\JetBrains\PyCharm Community Edition 2019.1.3\helpers\pydev\pydevd.py", line 1752, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "D:\Apps\JetBrains\PyCharm Community Edition 2019.1.3\helpers\pydev\pydevd.py", line 1147, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "D:/Multiverse_Projects/Final year project- Fall Detection/video-fall-detection/cloud-firestore-test.py", line 28, in <module>
    main()
  File "D:/Multiverse_Projects/Final year project- Fall Detection/video-fall-detection/cloud-firestore-test.py", line 22, in main
    u'born': 1815
  File "D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\google\cloud\firestore_v1\document.py", line 234, in set
    write_results = batch.commit()
  File "D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\google\cloud\firestore_v1\batch.py", line 147, in commit
    metadata=self._client._rpc_metadata,
  File "D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\google\cloud\firestore_v1\gapic\firestore_client.py", line 1042, in commit
    request, retry=retry, timeout=timeout, metadata=metadata
  File "D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\google\api_core\gapic_v1\method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\google\api_core\retry.py", line 286, in retry_wrapped_func
    on_error=on_error,
  File "D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\google\api_core\retry.py", line 184, in retry_target
    return target()
  File "D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\google\api_core\timeout.py", line 214, in func_with_timeout
    return func(*args, **kwargs)
  File "D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\google\api_core\grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "D:\Multiverse_Projects\Final year project- Fall Detection\video-fall-detection\environment\lib\site-packages\six.py", line 740, in raise_from
    raise value
google.api_core.exceptions.ServiceUnavailable: 503 Deadline Exceeded

Process finished with exit code 1

```

Thanks!

**Comment by HemangChothani** [comment]
@praharshbhatt  Can you please try with the upgraded version of `google-cloud-firestore` and it's dependencies.

**Comment by kaiyiuw** [comment]
I got same error just these days. 
Code below can write to firestore from my MacBookPro and from (bare) Alibaba-Cloud machine. 

However, when this code is inside a django+nginx project (in Alibaba-Cloud), it will hang in the last line of code (db_ref.set) and finally, 503 deadline exceeded. 

I did upgraded all package related to google-cloud or firebase, but issue still there. 

#-------------------------------
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore
import os
import datetime
script_dir = os.path.dirname(__file__)

def saveTrendyNewsFromWebView(userEmail, newsTitle):
    if (not len(firebase_admin._apps)):
        cred = credentials.Certificate(os.path.join(script_dir, 'xxx.json'))
        firebase_admin.initialize_app(cred)
    db = firestore.client()
    
    today_str = datetime.date.today().strftime("%Y.%m.%d")
    time_str = datetime.datetime.now().strftime("%H:%M:%S")
    
    doc_ref = 
               db.collection('User').document(userEmail).collection('News').document(today_str)
   
     doc_ref.set({
        'recordArray': firestore.ArrayUnion([{'newsTitle': newsTitle, 'Date': time_str}])
    }, merge=True)

#-------------------------------

**Comment by HebeHH** [comment]
I'm also getting this error, for `client.document_text_detection(image=image, timeout=60)`. Some of the confusing aspects:

1. It is very, very intermittent. I'm making several successive calls, ~30sec apart. Sometimes, it will throw this error for all of them. Other times, it's fine. I'm not sure if I'm overloading the API or what? It's somewhat sketch if so, since [I should apparently be able to make a thousand or so a minute.](https://cloud.google.com/vision/pricing) 
2. The error is _incredibly confusing._ **_INCREDIBLY._** Why the emphasis? Because the status code does not match the error. According to [Google's documentation](https://cloud.google.com/apis/design/errors), the `503` status code means `UNAVAILABLE`. Meanwhile, the `DEADLINE EXCEEDED` status message is meant to be accompanied by the `504` status code. According to Google, these are two separate statuses with two different causes. If it's `503 UNAVAILABLE`, it's an issue with Google's server that we can't do anything about. But if it's `504 DEADLINE EXCEEDED`, then the user needs to give a longer deadline. Would be nice to know which.

Given that I'm the only one using a `timeout`, I suspect the correct error should be `503 UNAVAILBLE`. But not sure. 


Tldr: I'm guessing this error means that Google's server is down, but unsure.

**Comment by HemangChothani** [comment]
@praharshbhatt @kaiyiuw @HebeHH I have tried to reproduce this error by some custom changes and i got  error `DeadlineExceeded: 504 Deadline Exceeded`,  not able to reproduce `ServiceUnavailable: 503 Deadline Exceeded` .

**Tested environment details:**

    google-cloud-firestore==1.7.0
    google-api-core==1.20.0
    google-cloud-core==1.3.0 

**Stack trace**

    /.local/lib/python3.6/site-packages/google/api_core/timeout.py in func_with_timeout(*args, **kwargs)
    212             """Wrapped function that adds timeout."""
    213             kwargs["timeout"] = next(timeouts)
    --> 214             return func(*args, **kwargs)
    215 
    216         return func_with_timeout

    /.local/lib/python3.6/site-packages/google/api_core/grpc_helpers.pyc in error_remapped_callable(*args, **kwargs)
     57             return callable_(*args, **kwargs)
     58         except grpc.RpcError as exc:
    ---> 59             six.raise_from(exceptions.from_grpc_error(exc), exc)
     60 home/home/mahadevmahadev
     61     return error_remapped_callable

    /.local/lib/python3.6/site-packages/six.pyc in raise_from(value, from_value)
    736 else:
    737     def raise_from(value, from_value):
    --> 738         raise value 

    DeadlineExceeded: 504 Deadline Exceeded

If it's still reproducible then it might be backend issue.

**Comment by crwilcox** [comment]
Unfortunately we haven't been able to reproduce this. However, the code is correct as well as your understanding. 

What I think is happening is the server has a child RPC that is reaching a DEADLINE. the detail you are seeing is from that exception. So the error code is correct, but the details may be a bit misleading, though it is trying to be more informative.

Hopefully that makes sense.
